### PR TITLE
Up to date pipeline - version 1

### PR DESCRIPTION
**Pull request by maiste**
This PR introduces the pipeline used to keep the documentation up to date on the ocurrent.org website. The goal is to centralise the documentation from the different files in the `ocurrent` organisation and, merge it with `occurent.org` skeleton.

It relies on git to keep the data up to date by pushing to a specific Github repository (in our case, it should be `occurent.github.io`). The configuration is written in YAML and is loaded by the pipeline.

Fixes another part of #7.

Notes:
 - [x] Need #8 to be merge
 - [x] Fix the issue with the YAML file loading.
 - [x] Support Authentification using the `Gihtub.Webhooks`
 - [x] Add a better explanation to the README and the `pipeline.mli` file.

### CODE REVIEW

File: bin/pipeline.ml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  in
+  let commit_id = Current.map Gh.Api.Commit.id head in
+  (repo.name, Git.fetch commit_id)
+
+let fetch_selections ~github ~repos =
+  let f selections (repo, files) =
+    let repo, commit = fetch_commit ~github ~repo () in
+    { repo; commit; files } :: selections
+  in
+  List.fold_left f [] repos
```
Comment (maiste) [inline bin/pipeline.ml:27]: Can be replaced with a `List.map` (@MisterDA)

File: bin/conf.ml [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        let indexes = index_of_json indexes in
+        { output; repos; indexes }
+    | _ -> failwith "Corrupted data - conf: this is not supposed to happen."
+end
+
+let github_remote conf =
+  let { Current_github.Repo_id.owner; Current_github.Repo_id.name }, branch =
+    conf.output
+  in
+  let url = Format.sprintf "git@github.com:%s/%s.git" owner name in
```
Comment (MisterDA) [inline bin/conf.ml:94]: Use `Printf.sprintf` rather than format for simple string interpolation.

File: bin/hugo.ml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    let digest t = Yojson.Safe.to_string (to_json t)
+    let pp f t = Fmt.pf f "%s" (Current_git.Commit.hash t.commit)
+  end
+
+  module Value = Current.Unit
+
+  let folder_generator job path =
+    let dir = fst (Fpath.split_base path) in
+    Logs.debug (fun log -> log "Check  folder [%s]" (Fpath.to_string dir));
```
Comment (MisterDA) [inline bin/hugo.ml:31]: You can use `"%a" Fpath.pp dir`.

File: bin/hugo.ml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    let f x = Yojson.Safe.to_string (File.Copy.to_yojson x) in
+    List.map f files
+  in
+  let indexes =
+    let f x = Yojson.Safe.to_string (File.Index.to_json x) in
+    List.map f indexes
+  in
+  let to_digest = String.concat "," (files @ indexes) in
+  let digest = Digest.string to_digest |> Digest.to_hex in
+  Logs.info (fun l -> l "Digest file to get hash: %s" digest);
```
Comment (MisterDA) [inline bin/hugo.ml:108]: _File digest_?

File: bin/main.ml [hunk 5]
```diff
@@ -0,0 +1,66 @@
+module Gh = Current_github
+
+let setup_log default_level = Prometheus_unix.Logging.init ?default_level ()
+let program_name = "pipeline"
+
+let has_role user = function
+  | `Viewer | `Monitor -> true
+  | `Builder | `Admin -> (
+      match Option.map Current_web.User.id user with
+      | Some ("github:maiste" | "github:tmcgilchrist") -> true
```
Comment (MisterDA) [inline bin/main.ml:10]: me me me
I want in

File: bin/main.ml [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      v ?authn ~has_role ~secure_cookies ~name:program_name routes)
+  in
+  Lwt_main.run
+    (Lwt.choose [ Current.Engine.thread engine; Current_web.run ~mode site ])
+
+open Cmdliner
+
+let setup_log =
+  let docs = Manpage.s_common_options in
+  Term.(const setup_log $ Logs_cli.level ~docs ())
```
Comment (MisterDA) [inline bin/main.ml:45]: You can also use and setup fmt.cli.

File: bin/utils.mli [hunk 7]
```diff
@@ -0,0 +1,65 @@
+module Cmd : sig
+  (** This module is a wrapper for command you would expect to execute from the
+      command line. *)
+
+  val copy_all :
+    ?cwd:Fpath.t ->
+    job:Current.Job.t ->
+    Fpath.t ->
+    Fpath.t ->
+    (unit, Rresult.R.msg) Lwt_result.t
```
Comment (MisterDA) [inline bin/utils.mli:10]: I'd suggest ditching rresult completely.

File: dune-project [hunk 8]
```diff
@@ -0,0 +1,30 @@
+(lang dune 3.3)
+(generate_opam_files true)
+
+(name ocurrentorg)
+(version 0.1.0)
+
+(package
+  (allow_empty)
```
Comment (MisterDA) [inline dune-project:8]: You can install your executable in the package.

**Review by MisterDA — APPROVED** [review]
Looks good to me! Just a bunch of minor suggestions. You're an awesome coder.

**Pull request merged**